File: library/evolution/src/main/java/cz/matfyz/evolution/schema/SMO.java
package cz.matfyz.evolution.schema;

import cz.matfyz.core.schema.SchemaCategory;
import cz.matfyz.evolution.exception.EvolutionException;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
    @JsonSubTypes.Type(value = CreateObject.class, name = "createObject"),
    @JsonSubTypes.Type(value = DeleteObject.class, name = "deleteObject"),
    @JsonSubTypes.Type(value = UpdateObject.class, name = "updateObject"),
    @JsonSubTypes.Type(value = CreateMorphism.class, name = "createMorphism"),
    @JsonSubTypes.Type(value = DeleteMorphism.class, name = "deleteMorphism"),
    @JsonSubTypes.Type(value = UpdateMorphism.class, name = "updateMorphism"),
    @JsonSubTypes.Type(value = Composite.class, name = "composite"),
})
public interface SMO {

    void up(SchemaCategory schema) throws EvolutionException;

    void down(SchemaCategory schema) throws EvolutionException;

    <T> T accept(SchemaEvolutionVisitor<T> visitor) throws EvolutionException;

}

File: client/src/types/schema/operation/smo.ts
import type { Category } from '../Category';

export enum SMOType {
    CreateObjex = 'createObject',
    DeleteObjex = 'deleteObject',
    UpdateObjex = 'updateObject',
    CreateMorphism = 'createMorphism',
    DeleteMorphism = 'deleteMorphism',
    UpdateMorphism = 'updateMorphism',
    Composite = 'composite',
}

export type SMOFromServer<T extends SMOType = SMOType> = {
    type: T;
};

export type SMO<T extends SMOType = SMOType> = {
    readonly type: T;
    toServer(): SMOFromServer<T>;
    up(category: Category): void;
    down(category: Category): void;
}
